asciidoc-confluence-publisher-converter/src/main/java/org/sahli/asciidoc/confluence/publisher/converter/AsciidocConfluenceConverter.java

```diff
@@ -56,58 +53,124 @@
+            createDirectories(assetsRootFolder);
 
-        ConfluencePublisherMetadata confluencePublisherMetadata = initializeConfluencePublisherMetadata(spaceKey, ancestorId);
+            extractTemplatesFromClassPathTo(templatesRootFolder);
 
-        AdocFileVisitor visitor = new AdocFileVisitor(asciidocRootFolderPath, generatedDocOutputPath, asciidocConfluenceTemplatesPath);
-        walkFileTree(Paths.get(asciidocRootFolderPath), visitor);
+            AsciidocPagesStructureProvider.AsciidocPagesStructure structure = asciidocPagesStructureProvider.structure();
+            List<AsciidocPage> asciidocPages = structure.pages();
+            List<ConfluencePageMetadata> confluencePages = buildPageTree(templatesRootFolder, assetsRootFolder, asciidocPages);
 
-        Map<String, List<ConfluencePageMetadata>> confluencePublisherMetadataRegistry = visitor.confluencePageMetadataRegistry();
+            ConfluencePublisherMetadata confluencePublisherMetadata = new ConfluencePublisherMetadata();
+            confluencePublisherMetadata.setSpaceKey(this.spaceKey);
+            confluencePublisherMetadata.setAncestorId(this.ancestorId);
+            confluencePublisherMetadata.setPages(confluencePages);
 
-        List<ConfluencePageMetadata> rootPages = confluencePublisherMetadataRegistry.get(generatedDocOutputPath);
-        if (rootPages != null) {
-            confluencePublisherMetadata.getPages().addAll(rootPages);
-            buildPageTree(confluencePublisherMetadata.getPages(), confluencePublisherMetadataRegistry, generatedDocOutputPath);
+            return confluencePublisherMetadata;
+        } catch (Exception e) {
+            throw new RuntimeException("Could not convert asciidoc pages", e);
         }
+    }
+
+    private static List<ConfluencePageMetadata> buildPageTree(Path templatesRootFolder, Path assetsRootFolder, List<AsciidocPage> asciidocPages) {
+        List<ConfluencePageMetadata> confluencePages = new ArrayList<>();
+
+        asciidocPages.forEach((asciidocPage) -> {
+            try {
+                Path pageAssetsFolder = determinePageAssetsFolder(assetsRootFolder, asciidocPage);
+                createDirectories(pageAssetsFolder);
+
+                AsciidocConfluencePage asciidocConfluencePage = newAsciidocConfluencePage(asciidocPage, templatesRootFolder, pageAssetsFolder);
+                Path contentFileTargetPath = writeToTargetStructure(asciidocPage, pageAssetsFolder, asciidocConfluencePage);
+
+                List<AttachmentMetadata> attachments = buildAttachments(asciidocPage, pageAssetsFolder, asciidocConfluencePage.attachments());
+                copyAttachmentsAvailableInSourceStructureToTargetStructure(attachments);
+
+                List<ConfluencePageMetadata> childConfluencePages = buildPageTree(templatesRootFolder, assetsRootFolder, asciidocPage.children());
+
+                ConfluencePageMetadata confluencePageMetadata = new ConfluencePageMetadata();
```

Comment:
A method for the creation of the the `ConfluencePageMetadata` would make this `try` block a little slicker.

Reply:
Fixed in f409276c01f91a2fa1f4f07d9ec071ac82bc7683